### ISSUE TITLE
`md-sidenav` and `md-content` areas do not scroll independently on FireFox

### ISSUE THREAD

**Issue by johnwebbcole** [angular/material#2454]
I have a layout with a `md-sidenav` and a `md-content` area.  On Chrome/Safari, these scroll independently, however, on FireFox, the entire page scrolls as one page.

I have a [codepen](http://codepen.io/johnwebbcole/pen/aOoXRL) (this is the same codepen as #2453) showing this.

Thx,

J

**Comment by tarkeshwars** [comment]
Is there a temporary fix available around this issue?

**Comment by Ontic-Entity** [comment]
Hello, 

Please use '<md-content layout="row" flex>' instead of '<div layout="row" flex>'

**Comment by tarkeshwars** [comment]
I was using 'md-content layout="column" flex' earlier. I changed it to 'md-content="row" flex' like you suggested. That still doesn't work.

**Comment by Ontic-Entity** [comment]
tarkeshwars,

A personal code example could bring more light to your specific issue.

**Comment by csvan** [comment]
Confirming that this does not work on Chrome, either. Same code pen as above.

Internally, my team has managed to get this to work by browser-specific hacks. For this CP for example, you can get it working on Chrome and FF by using:

```
html, body { height: 100%; }
```

And then on Chrome:

```
md-content { height: 100%; }
```

and FF:

```
md-sidenav { height: 90vh; }
```

Of course, this is far from optimal. The problem in all cases seems to be that md-content does not properly force its children to conform to the height constraint of the parent element. Everything simply flows all over the place.

**Comment by dmastag** [comment]
Well is this what you are trying to achieve ?
http://codepen.io/dmastag/pen/oXORpZ

I changed the div to md-content as @Ontic-Entity suggested.

Before:
![image](https://cloud.githubusercontent.com/assets/5745805/9287318/d7e7a632-4338-11e5-92ca-be4dd776da2c.png)

After:
![image](https://cloud.githubusercontent.com/assets/5745805/9287319/e66739fc-4338-11e5-8d24-ddd7f2d4f947.png)

As I understand it, you will need to use an md-content if you want the content to be scrollable.

**Comment by tarkeshwars** [comment]
Hey @dmastag  The fix you suggested works on both firefox and chrome but after adding it the sidenav on smaller screen doesn't cover the toolbar when it opens!

**Comment by dmastag** [comment]
@tarkeshwars could you explain the meaning of
> the sidenav on smaller screen doesn't cover the toolbar when it opens

Looks like it is working for me (Chrome Version 44.0.2403.155 m)

![image](https://cloud.githubusercontent.com/assets/5745805/9360607/4a9be824-46c1-11e5-8ffa-87bc35d34448.png)

![image](https://cloud.githubusercontent.com/assets/5745805/9360601/45dd9d8c-46c1-11e5-8aae-9df8c51de029.png)

**Comment by tarkeshwars** [comment]
This is surprising, Even when I open you codepen on my mac+chrome, I see this: 
![screen shot 2015-08-19 at 10 03 28 pm](https://cloud.githubusercontent.com/assets/10356173/9362489/482b7b0c-46be-11e5-8d0e-482be1308bb0.png)

I check on Firefox as well, it looks the same way there as well.

**Comment by tarkeshwars** [comment]
Just cross checked on chrome+windows 7, it looks similar even there!

**Comment by dmastag** [comment]
I have no idea how this happened, but I just looked at my codepen and it happens just like you said.
Why is it behaving differently? I have no idea.

![image](https://cloud.githubusercontent.com/assets/5745805/9366114/91c0b75c-46df-11e5-98ca-41911696bff0.png)

**Comment by tarkeshwars** [comment]
Your answer to keeping things scroll in parallel was right. The problem we are seeing right now is because of md-sidenav taking height of 100%. When it's inside a div it takes the full window height whereas if its inside an md-content tag then it takes the height of md-content tag. But I don't understand why it is treating a div and md-content so differently.

**Comment by topherfangio** [comment]
Sorry for jumping in the conversation so late; I will be looking at this issue over the next few days to try to get it resolved.

Regarding the difference between div/md-content; md-content sets `position: relative` whereas a standard div doesn't, so that is causing the difference in toolbar overlap and also probably the scrolling issues.

I'll work on it, but please let me know if you find anything else that may help :-)

**Comment by tarkeshwars** [comment]
I ended up taking the approach suggested by @dmastag and added a bit of css hack. This could be a solution for those looking for a temporary fix My layout is like this:
```html   
 <body layout="column">
     <div>
        <!--I had a toolbar and tabs in it -->
     </div>
    <md-content layout="row" id="parentcontainer">
       <md-sidenav>
       </md-sidenav>
       <md-content>
       </md-content>
    </md-content>
</body>
 ```
I then gave a negative margin for parentcontainer and added an equal padding to md-sidenav.md-locked-open and the sibling md-content. This works like a charm. Its tested in latest versions Chrome, Firefox, Opera and Vivaldi and Safari on both Mac and Windows. Here's a working demo of it in a webapp that I am building:
www.fundamine.com

**Comment by dmastag** [comment]
@tarkeshwars your site looks great. Glad that it is finally working out for you

**Comment by tarkeshwars** [comment]
Thanks.

**Comment by topherfangio** [comment]
Based on some feedback with the team, I think the following is the ideal structure that we should support, however, this is not currently working on Firefox as the entire content scrolls instead of just the `<md-content>` portion.

I will continue to investigate this.

http://codepen.io/topherfangio/pen/LpVWVp

**Comment by ThomasBurleson** [comment]
@dmastag - I updated your CodePen to use ngMaterial 0.11.2 and Angular 1.3.19

http://codepen.io/ThomasBurleson/pen/LpWVxJ

> To expedite resolution, please always test with the latest versions:
*  Angular Material HEAD
  *  Bower Install - https://github.com/angular/bower-material/blob/master/angular-material.js
  *  Bower Install - https://github.com/angular/bower-material/blob/master/angular-material.css 
  *  RAWGIT - https://rawgit.com/angular/bower-material/master/angular-material.js
  *  RAWGIT - https://rawgit.com/angular/bower-material/master/angular-material.css
* AngularJS Latest Release versions:
  *  Angular 1.3.19 (or greater)
  *  Angular 1.4.6 (or greater).

**Comment by topherfangio** [comment]
@ThomasBurleson With the structure you provided, the sidenav will appear underneath the toolbar when opened on a small screen because the `<md-content>` tag sets `position: relative` (see image in [comment](https://github.com/angular/material/issues/2454#issuecomment-132744165)) and the sidenav uses `position: absolute`.

If you use a plain `<div>` instead of the `<md-content>`, then the sidenav is correctly positioned, but scrolling then causes the toolbar to scroll too because the `<div>`'s height is not calculated properly by the browser (Firefox specifically). Setting the `<div>` to `overflow: hidden` fixes this, but having to add this is somewhat a pain and not automatic/easy.

I've updated my [Codepen](http://codepen.io/topherfangio/pen/gampyP) to show what I mean (and use 11.2). If you make it small enough so the sidenav is closed, you can open it and see how it properly appears on top of the toolbar, but the content area scrolls the toolbar as well. You can uncomment the class in the CSS to fix the issue with the `overflow: hidden` trick.

_Last Note: I tried the `height: inherit` trick from the other issue (#4022), but it doesn't cause Firefox to calculate the height for some reason, so it doesn't appear to be a viable option here._

**Comment by germanferrero** [comment]
I have many issues with md-content height / scroll in firefox. Right now it's making my app unusable in this browser. I will update with some plunkers in my free time. In few words, like examples shown here, md-content doesn't srcoll independently.I have fab-speed-dial in botom-right corner which it's taking out of screen, and must scroll to see it. Really need this get fixed. I desired priority get critical again.

**Comment by topherfangio** [comment]
@germanferrero I have been looking at this over the last two days and think I may have a more suitable solution.

Can you try using the latest version and adding the following CSS to see if it alleviates the issue?

```css
.layout-column.flex, .layout-row.flex {
    overflow: auto;
}
```

This seems to fix the Firefox-specific issue mentioned here, and I believe it makes logical sense, but may have some unintended side-effects. If you see any major issues, could you please post back here so we can see how much this would change things?

**Comment by germanferrero** [comment]
@topherfangio, this is the moment when everyone claps! Im loving those three lines.
Obviously, like you said, they bring whit them some side-effects. My examples were:

Cases where i need a layout-row.flex element with a child that exceed  parent edges, something like this:
![o9mhp](https://cloud.githubusercontent.com/assets/7140856/10472426/aadc0484-71f6-11e5-8be2-505d04cb7757.png)
 White div is a child of big-white div with high z-index and top:-70px to break edge.
But in this case i only need to override overflow with overflow:visible.

Other (many) cases where i have  whiteframe-zX elements, shadow get cut almost always.

And some little things that were working "thanks" to this bad behavior must be fixed.

That's all for now!  Three magic lines are staying in my code definitely. Thanks you so much, and I will let you know if something wrong occurs.

**Comment by topherfangio** [comment]
@germanferrero I'm working on a PR that will hopefully have an even better fix so that you don't have to worry quite so much about the whiteframes and other side-effects. Should be ready soon ;-)

**Comment by naomiblack** [comment]
This is a browser-specific issue that we are unfortunately unable to resolve for 1.0. However, we will document CSS best practices for working around this type of issue.

**Comment by topherfangio** [comment]
After discussion with the team, we have decided to close this in favor of #5652 which will help document workarounds to common layout issues. We will revisit this (and similar issues) after some layout modifications that we are planning post 1.0.

**Comment by sombriks** [comment]
i stumbled on this very same issue and what worked for me was to use an md-list inside md-content.

i am unable to provide a sample, but it`s working well even on ios devices, which are the most problematic ones with angular material.

**Comment by madhu131313** [comment]
I simply added `position:fixed` for the sidenav and it's working fine

**Comment by gombosg** [comment]
Still not fixed as of 2019. Only Chrome is affected. The `height: 100%` hack worked for me.

**Comment by Splaktar** [comment]
@gombosg if you are seeing a specific issue and have a reproduction, please create a new issue.
